Mathlib/RingTheory/MvPolynomial/NewtonIdentities.lean

```diff
@@ -0,0 +1,325 @@
+Authors: Michael Lee
+-/
+import Mathlib.Algebra.Algebra.Subalgebra.Basic
+import Mathlib.Data.Finset.Card
+import Mathlib.Data.Fintype.Basic
+import Mathlib.Data.MvPolynomial.CommRing
+import Mathlib.Data.MvPolynomial.Rename
+import Mathlib.Data.Nat.Parity
+import Mathlib.RingTheory.MvPolynomial.Basic
+import Mathlib.RingTheory.MvPolynomial.Symmetric
+import Mathlib.RingTheory.Polynomial.Basic
+
+/-!
+# Newton's Identities
+
+This file defines `MvPolynomial` power sums as a means of implementing Newton's identities.
+
+## Main declarations
+
+* `MvPolynomial.psum`
+
+* `MvPolynomial.esymm_recurrence` encapsulating the primary result
+
+## Notation
+
++ `psum σ R n` is the degree-`n` power sum in `MvPolynomial σ R`.
+
+As in other polynomial files, we typically use the notation:
+
++ `σ τ : Type _` (indexing the variables)
+
++ `R S : Type _` `[CommSemiring R]` `[CommSemiring S]` (the coefficients)
+
++ `r : R` elements of the coefficient ring
+
++ `i : σ`, with corresponding monomial `X i`, often denoted `X_i` by mathematicians
+
++ `φ ψ : MvPolynomial σ R`
+
+## References
+
+See [zeilberger1984] for the combinatorial proof of Newton's identities.
+-/
+
+open Equiv (Perm)
+
+open BigOperators MvPolynomial
+
+noncomputable section
+
+section PowerSum
+
+namespace Multiset
+
+variable {R : Type _} [CommSemiring R]
+
+/-- The degree-`n` power sum evaluated at the elements of `s` -/
+def psum (s : Multiset R) (n : ℕ) : R := (s.map (fun (x : R) ↦ x ^ n)).sum
+
+end Multiset
+
+open Finset
+
+variable (σ R) [CommSemiring R] [Fintype σ] [Fintype τ]
+
+/-- The degree-`n` power sum -/
+def psum (n : ℕ) : MvPolynomial σ R := ∑ i, (X i) ^ n
+
+theorem rename_psum (n : ℕ) (e : σ ≃ τ) : rename e (psum σ R n) = psum τ R n := by
+  calc
+    rename e (psum σ R n) = ∑ i, (X (e i)) ^ n := by
+      simp_rw [psum, map_sum, map_pow, rename_X]
+    _ = ∑ i in univ.map e.toEmbedding, (X i) ^ n := by simp [-map_univ_equiv]
+    _ = ∑ i, (X i) ^ n := by rw [map_univ_equiv]
+
+theorem psum_isSymmetric (n : ℕ) : IsSymmetric (psum σ R n) := by
+  intro
+  rw [rename_psum]
+
+end PowerSum
+
+namespace Newton
+
+open Finset Nat
+
+variable (σ : Type _) [Fintype σ] [DecidableEq σ] [Fintype τ] (R : Type _) [CommRing R]
+  [NoZeroDivisors R] [CharZero R]
+
+-- The following proof is from Zeilberger, "A combinatorial proof of Newton's identities" (1984)
+def PairsPred (k : ℕ) (t : Finset σ × σ) : Prop := card t.fst ≤ k ∧
+  ((card t.fst = k) → t.snd ∈ t.fst)
+
+instance : DecidablePred (PairsPred σ k) := Classical.decPred (PairsPred σ k)
+
+def pairs (σ : Type _) [Fintype σ] (k : ℕ) : Finset (Finset σ × σ) :=
+  Finset.univ.filter (PairsPred σ k)
+
+def weight (k : ℕ) (t : Finset σ × σ) : MvPolynomial σ R :=
+  (-1) ^ (card t.fst) * ((∏ a in t.fst, X a) * (X t.snd) ^ (k - card t.fst) : MvPolynomial σ R)
+
+def T_map (t : Finset σ × σ) : Finset σ × σ :=
+  if t.snd ∈ t.fst then (t.fst.erase t.snd, t.snd) else (insert t.snd t.fst, t.snd)
```

Comment:
I think this might be slightly easier to work with as
```suggestion
  if h : t.snd ∈ t.fst then (t.fst.erase t.snd, t.snd) else (cons t.snd t.fst h, t.snd)
```

Reply:
Done!